src/serialize.rs

```diff
@@ -0,0 +1,311 @@
+use core::fmt;
+
+use alloc::{format, string::ToString, vec::Vec};
+use serde::{Deserialize, Serialize};
+
+use crate::{Block, Function, Inst, InstRange, MachineEnv, Operand, PRegSet, RegClass, VReg};
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
+enum InstOpcode {
+    Op,
+    Ret,
+    Branch,
+}
+
+impl fmt::Display for InstOpcode {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            InstOpcode::Op => f.write_str("op"),
+            InstOpcode::Ret => f.write_str("ret"),
+            InstOpcode::Branch => f.write_str("branch"),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+struct InstData {
+    op: InstOpcode,
+    operands: Vec<Operand>,
+    clobbers: PRegSet,
+    is_safepoint: bool,
+}
+
+/// A wrapper around a `Function` and `MachineEnv` that can be serialized and
+/// deserialized.
+///
+/// The serialized form of this structure is not stable: it is intended to be
+/// deserialized with the exact same version of regalloc2 as the one that it
+/// was created with.
+#[derive(Serialize, Deserialize)]
+pub struct SerializableFunction {
+    machine_env: MachineEnv,
+    entry_block: Block,
+    insts: Vec<InstData>,
+    blocks: Vec<InstRange>,
+    block_preds: Vec<Vec<Block>>,
+    block_succs: Vec<Vec<Block>>,
+    block_params_in: Vec<Vec<VReg>>,
+    block_params_out: Vec<Vec<Vec<VReg>>>,
+    num_vregs: usize,
+    reftype_vregs: Vec<VReg>,
+    debug_value_labels: Vec<(VReg, Inst, Inst, u32)>,
+    spillslot_size: Vec<usize>,
+    multi_spillslot_named_by_last_slot: bool,
+    allow_multiple_vreg_defs: bool,
+}
+
+impl SerializableFunction {
+    /// Creates a new `SerializableFunction` from an arbitray `Function` and
+    /// `MachineEnv`.
+    pub fn new(func: &impl Function, machine_env: MachineEnv) -> Self {
```

Comment:
Should the second arg be a `&MachineEnv`? Alternately, no clone two lines below?

Reply:
Removed the clone.